source/byte_buf.c

```diff
@@ -380,6 +380,75 @@ int aws_byte_buf_append(struct aws_byte_buf *to, const struct aws_byte_cursor *f
     return AWS_OP_SUCCESS;
 }
 
+int aws_byte_buf_append_dynamic(struct aws_byte_buf *to, const struct aws_byte_cursor *from) {
+    assert(from->ptr);
+    assert(to->buffer);
+
+    if (to->capacity - to->len < from->len) {
+        /*
+         * We don't have size_t math so use 64 bits for overflow checking.
+         * Then do a final check before allocating to verify that the computed
+         * new capacity is actually containable within the platform's size_t type.
+         *
+         * If size_t ever exceeds 64 bits, this won't be valid
+         */
+
+        /*
+         * NewCapacity = Max(OldCapacity * 1.5, OldCapacity + MissingCapacity)
+         */
+        uint64_t missing_capacity = (uint64_t)(from->len - (to->capacity - to->len));
+        uint64_t new_capacity = 0;
+        if (aws_add_u64_checked(to->capacity, missing_capacity, &new_capacity)) {
+            return AWS_OP_ERR;
+        }
+
+        uint64_t growth_capacity = 0;
+        if (aws_add_u64_checked(to->capacity, to->capacity / 2, &growth_capacity)) {
+            return AWS_OP_ERR;
```

Comment:
if growth_capacity overflows, we could still try using the one based on missing_capacity